design/20220118.certificate-issuance-exponential-backoff.md

```diff
@@ -0,0 +1,144 @@
+# Exponential backoff for certificate issuance
+
+## Summary
+
+Implement a way to apply exponential backoff when retrying failed certificate
+issuances. Issuance in this context refers to the period of time during which
+the `Issuing` condition on a `Certificate` is set to true and for which a new
+set of issuance-specific resources (`CertificateRequest`s, `Order`s, `Challenge`s etc) is
+created.
+
+## Motivation
+
+Currently failed issuances are retried once an hour without a backoff or time limit. This means that 1) continuous failures in large installations can overwhelm external services 2) in case of some external problems (i.e DNS) that cause issuances to fail and last for longer than a couple of hours, rate limits (see [Let'sEncrypt rate limts](https://letsencrypt.org/docs/rate-limits/)) can be easily hit.
+
+## Goals
+
+- Ensure that retrying failed issuances does not overwhelm external services and is less likely to cause issuance being rate-limited by backing off with an exponentially increasing time period. 
+
+- Ensure that when the backoff is being applied, users have a way to find out when the issuance will be next retried and that the backoff mechanism does not introduce extra complexity for debugging and fits in with the already existing 1 hour backoffs
```

Comment:
Hm - I recall you talked about setting `nextIssuanceTime` to the next time the renewal attempt will be made on Slack and we eventually thought it'd be best to not do that.

I *don't* see that discussed in the proposal, but this is still a goal (and I think an un-met one with this proposal as-is)? Unless I missed something 😄

Reply:
I still wanted to give users some way how to tell when the next issuance will be attempted (this type of question seems to appear a lot, so it seems like something that would be useful for their debugging) beyond just telling them to calculate it.

For this design, I was aiming to fulfill that goal with just an event on the `Certificate`, I will add that to the spec 👍🏼 (I am still not entirely sure what would be the best way to expose this to users, but I guess a status field could always be add later if there was a  need for it)

Reply:
As I was working on the implementation, the event didn't seem all that useful- it is easy to have multiple of these events with different timestamps at the same time time on one failing `Certificate` when manually retrying issuances which is confusing, also the fact that the events are short lived may not help much with debugging.